typed/helpers.go

```diff
@@ -114,17 +114,27 @@ func resolveSchema(s *schema.Schema, tr schema.TypeRef, v *value.Value, ah atomH
 	return handleAtom(a, tr, ah)
 }
 
-func deduceAtom(a schema.Atom, v *value.Value) schema.Atom {
+// deduceAtom determines which of the possible types in atom 'atom' applies to value 'val'.
```

Comment:
I don't understand this, atoms only permit one type?

Reply:
We changed that to allow for a generic schema for CRDs

https://github.com/kubernetes-sigs/structured-merge-diff/pull/73

Reply:
I don't see how that's better than untyped but I'll take your word for it for the moment.

Reply:
The reason we did it this way was, to be able to clearly define the behavior of different Untyped elementRelationships, like "atomic" and "separable". We wanted different a elementRelationship and elementType for lists and maps in what would have been the "untyped separable" case (atomic lists and separable maps), and we wanted to solve it in a generic way so we don't have to duplicate the logic and don't have to hide the definition of its behavior in code.

 Also, as a side effect we should also be able to eventually use this for CRD validation shemas which define a field like "map or array of maps".